File: server/pkg/server/middleware/grpc.go
package middleware

import (
	"context"
	grpc_zap "github.com/grpc-ecosystem/go-grpc-middleware/logging/zap"
	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
	"go.uber.org/zap"
	"go.uber.org/zap/zapcore"
	"google.golang.org/genproto/googleapis/rpc/errdetails"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func LogMessageProducer() grpc_zap.MessageProducer {
	return func(ctx context.Context, msg string, level zapcore.Level, code codes.Code, err error, duration zapcore.Field) {

		fields := []zap.Field{
			zap.Error(err),
			zap.String("grpc.code", code.String()),
			duration,
		}

		st := status.Convert(err)
		for _, detail := range st.Details() {
			switch t := detail.(type) {
			case *errdetails.DebugInfo:
				fields = append(fields, zap.String("err.debug.detail", t.Detail))
				fields = append(fields, zap.Strings("err.debug.stack", t.StackEntries))
			case *errdetails.ErrorInfo:
				fields = append(fields, zap.String("err.error.reason", t.Reason))
			case *errdetails.BadRequest:
				for _, violation := range t.GetFieldViolations() {
					fields = append(fields, zap.String("err.request.field", violation.GetField()))
					fields = append(fields, zap.String("err.request.violation", violation.GetDescription()))
				}
			}
		}

		ctxzap.Extract(ctx).Check(level, msg).Write(
			fields...
		)
	}
}

File: client/src/app/lib/util/grpc.ts
import { HttpErrorResponse } from '@angular/common/http';

export function fieldErrors(err: HttpErrorResponse, fieldName: string): string[] {
  if (err === null || err === undefined) {
    return [];
  }
  const errors: string[] = [];
  if (err?.error?.details?.length > 0) {
    for (const v of err.error.details) {
      switch (v['@type']) {
        case 'type.googleapis.com/google.rpc.BadRequest':
          if (v?.field_violations?.length > 0) {
            for (const f of v.field_violations) {
              if (f.field === fieldName) {
                errors.push(f.description);
              }
            }
          }
          break;
      }
    }
  }
  return errors;
}
